### ISSUE TITLE
[explicit-function-return-type] should not lint react render function

### ISSUE THREAD

**Issue by XHMM** [typescript-eslint/typescript-eslint#492]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": ["error"]
  }
}
```

```TS
import * as React from 'react'
import * as ReactDOM from 'react-dom'

const Test = () => <div>hi</div>

/* // class style
class Test extends React.Component {
  render() {
    return <div>hi</div>
  }
}
*/

ReactDOM.render(<Test />, document.querySelector('#app'))
```

**Expected Result**
should not lint `() => <div>hi</div>` and render function

**Actual Result**
![image](https://user-images.githubusercontent.com/24786522/57061517-b91a4180-6cef-11e9-8dcd-0a2c1957e137.png)
![image](https://user-images.githubusercontent.com/24786522/57061678-5aa19300-6cf0-11e9-901a-4801a7727e43.png)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.7.0` |
| `@typescript-eslint/parser`        | `1.7.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `yarn`                              | `1.15.2` |

**Comment by bradzacher** [comment]
For your first example of `const Test = () => <div />;`, the rule is working as expected.
The rule enforces explicitly adding return types to all functions.
Both your variable + arrow function have no explicit type on them, so they fail the rule.

If you want this to work without typing, you must use the functional component type for your variable from the react typings.

---

As for your other case, we haven't added support for extending methods from base classes.

**Comment by XHMM** [comment]
Thanks for reply , I also tried add type to variable only, but it seems not work：

![image](https://user-images.githubusercontent.com/24786522/57091770-80ed2000-6d3c-11e9-8fd7-10307a1b3b55.png)

**Comment by bradzacher** [comment]
Check the docs for the rule.
There are settings you'll need to turn on to make it work.

**Comment by XHMM** [comment]
Thanks, I turned on `allowTypedFunctionExpressions` and it works

**Comment by menglingyu** [comment]
add : JSX.Element like 
```
 render(): JSX.Element {

```
